src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,272 @@
+import {
+  ExpectedVersionConflictError,
+  STREAM_DOES_NOT_EXIST,
+  assertExpectedVersionMatchesCurrent,
+  type EventStore,
+  type Event,
+  type AggregateStreamOptions,
+  type AggregateStreamResult,
+  type AppendToStreamOptions,
+  type AppendToStreamResult,
+  type ReadStreamOptions,
+  type ReadStreamResult,
+  type ReadEvent,
+  type ReadEventMetadata,
+  type ExpectedStreamVersion,
+} from '@event-driven-io/emmett';
+import { type Collection, MongoClient, type WithId } from 'mongodb';
+import { v4 as uuid } from 'uuid';
+
+export const MongoDBEventStoreDefaultStreamVersion = -1;
+export const MongoDBDefaultCollectionName = 'eventstreams';
+
+export type StreamType = string;
+export type StreamName<T extends StreamType = StreamType> = `${T}:${string}`;
+
+export type StreamNameParts<T extends StreamType = StreamType> = {
+  streamType: T;
+  entityId: string;
+};
+
+export type StreamToProject<EventType extends Event> = {
+  streamName: StreamName;
+  streamType: StreamType;
+  entityId: string;
+  streamVersion: number;
+  events: ReadEvent<EventType, ReadEventMetadata>[];
+};
+
+export type Projection<EventType extends Event> = (
+  stream: StreamToProject<EventType>,
+) => void | Promise<void>;
+
+export interface EventStream<EventType extends Event> {
+  streamName: string;
+  events: Array<ReadEvent<EventType, ReadEventMetadata>>;
+  createdAt: Date;
+  updatedAt: Date;
+}
+export type EventStreamEvent<EventType extends Event> =
+  EventStream<EventType>['events'][number];
+
+export interface MongoDBConnectionOptions {
+  connectionString: string;
+  database: string;
+  collection?: string;
+}
+
+class EventStoreClass implements EventStore<number> {
+  private readonly collection: Collection<EventStream<Event>>;
+
+  constructor(collection: typeof this.collection) {
```

Comment:
**QUESTION:** Do we assume for now that we will perform all operations on the single collection side? @alex-laycalvert did you think already about transactions or sessions handling? This could be useful for e.g. projections to other collections.

Reply:
I started this with the assumption that transactions aren't possible since my project doesn't use any replica sets. I'm definitely open to the idea of multiple collections based on stream type.

Reply:
I think that it's a safe choice to provide the initial capabilities without those additional features. We might still want to provide them through append options, but this can happen in a separate PR.